src/core/sock/sockinfo.h

```diff
@@ -644,4 +649,80 @@ class sockinfo : public socket_fd_api,
     //////////////////////////////////////////////////////////////////
 };
 
+typedef std::unordered_map<flow_tuple, bool> port_no_bind_map_t;
+struct port_map_list {
+    std::list<in_port_t>::iterator it;
+    port_no_bind_map_t map;
+};
+
+class bind_no_port {
+public:
+    /*  Logic of IP_BIND_ADDRESS_NO_PORT
+        In case we need a new port from OS, we call bind with port 0,
+        Otherwise - we call bind with specific port from our DB. */
+    int bind_and_set_port_map(const sock_addr &src, const sock_addr &dst, int fd)
+    {
+        int ret = 0;
+        sock_addr addr(src);
+        socklen_t addr_len = sizeof(addr);
+        std::lock_guard<decltype(m_lock)> lock(m_lock);
+
+        in_port_t reused_port = 0;
+        flow_tuple temp_mapper(dst.get_ip_addr(), dst.get_in_port(), src.get_ip_addr(),
+                               src.get_in_port(), PROTO_TCP, src.get_sa_family());
+        for (std::list<in_port_t>::iterator _port_it = m_port_list.begin();
+             _port_it != m_port_list.end(); ++_port_it) {
+            auto _iter = arr[*_port_it].map.find(temp_mapper);
+            if (_iter == arr[*_port_it].map.end()) {
+                reused_port = *_port_it;
+                arr[*_port_it].it = _port_it;
+                break;
+            }
+        }
+
+        if (reused_port) {
+            addr.set_in_port(reused_port);
+            arr[reused_port].map[temp_mapper] = true;
+        }
+
+        if ((ret = orig_os_api.bind(fd, addr.get_p_sa(), addr_len))) {
+            return ret;
+        }
+
+        if (!reused_port) {
+            if ((ret = orig_os_api.getsockname(fd, addr.get_p_sa(), &addr_len))) {
+                return ret;
+            }
+            in_port_t new_port = addr.get_in_port();
+            m_port_list.push_front(new_port);
+            arr[new_port].map[temp_mapper] = true;
+            arr[new_port].it = m_port_list.begin();
+        }
+        return ret;
+    }
+
+    void release_port(const sock_addr &src, const sock_addr &dst)
+    {
+        in_port_t port_to_release = src.get_in_port();
+        flow_tuple temp_mapper(dst.get_ip_addr(), dst.get_in_port(), src.get_ip_addr(),
+                               port_to_release, PROTO_TCP, src.get_sa_family());
+        temp_mapper.set_src_port(0);
+        std::lock_guard<decltype(m_lock)> lock(m_lock);
+
+        if (arr[port_to_release].map.find(temp_mapper) != arr[port_to_release].map.end()) {
+            arr[port_to_release].map.erase(temp_mapper);
+            if (arr[port_to_release].map.size() == 0) {
+                m_port_list.erase(arr[port_to_release].it);
+            }
+        }
+    }
+
+private:
+    lock_spin_recursive m_lock;
+    std::list<in_port_t> m_port_list;
+    struct port_map_list arr[65536];
```

Comment:
Please use MAX_UINT16

Reply:
please use m_arr

Reply:
Replacing arr with unordered_map<port, unordered_set<flow_tuple>> can simplify things i guess.  you can search quickly if port exists iand if tuple exist for that port and much less management of lists and iterators.
What do you think?